
Conectar drive personal

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Importar librerías necesarias e instalar modelo para modelado de tópicos BERTopic

In [ ]:
import pandas as pd
!pip install bertopic

Importar modelo BERTopic e instanciarlo con parámetros de configuración básicos: lenguaje, cálculo de probabilidad de los tópicos obtenidos por documento y seguimiento de las fases del modelo (verbose)

In [ ]:
#IMPORTAR MODELO
from bertopic import BERTopic
topic_model = BERTopic(language="english", calculate_probabilities=False, verbose=True)

Primero se lee el dataset con los tweets limpios y pre-procesados. A continuación, se transforma la columna 'body' con los tweets a una lista de strings para utilizarla como entrada al modelo BERTopic. Finalmente, se obtienen los tópicos más relevantes del tópico

In [ ]:
#leer dataset
data_tesla = pd.read_csv('/content/drive/MyDrive/TFG ADE/Data_Tesla_Limpio.csv')
data=(data_tesla['body'].astype(str)).tolist()
topics, _ = topic_model.fit_transform(data)

Batches:   0%|          | 0/5988 [00:00<?, ?it/s]

2023-11-14 17:43:30,801 - BERTopic - Transformed documents to Embeddings
2023-11-14 17:51:43,932 - BERTopic - Reduced dimensionality
2023-11-14 17:52:03,969 - BERTopic - Clustered reduced embeddings


Se obtiene información general sobre los tópicos obtenidos por el modelo: nombre/identificador, términos representativos y documentos representativos (-1 es el tópico outlier, no se tiene en cuenta en el análisis)

In [ ]:
freq = topic_model.get_topic_info()
freq

Topic  Count                                               Name  \
0        -1  85381                      -1_autopilot_musk_car_battery   
1         0   2116                        0_solar_roof_solarcity_tile   
2         1    992                              1_nan_nano_ching_yeah   
3         2    980                          2_gspc_surge_fall_outlook   
4         3    884                       3_elon_elons_stuttering_tapp   
...     ...    ...                                                ...   
2122   2121     10             2121_signal_issued_potnetial_outpretty   
2123   2122     10               2122_burger_sandwich_restocked_pbhto   
2124   2123     10  2123_localnational_fortuitous_extensionbased_b...   
2125   2124     10                     2124_4200_yuasa_jvor_whyteslas   
2126   2125     10             2125_expiry_expiring_11sold_sweepsjuly   

                                         Representation  \
0     [autopilot, musk, car, battery, elonmusk, sale...   
1     [solar, roof, solarcity, tile, city, panel, ba...   
2                  [nan, nano, ching, yeah, , , , , , ]   
3     [gspc, surge, fall, outlook, nflx, soar, rise,...   
4     [elon, elons, stuttering, tapp, lie, wright, g...   
...                                                 ...   
2122  [signal, issued, potnetial, outpretty, abux, n...   
2123  [burger, sandwich, restocked, pbhto, bintsla, ...   
2124  [localnational, fortuitous, extensionbased, bu...   
2125  [4200, yuasa, jvor, whyteslas, bydgf, jvaesc, ...   
2126  [expiry, expiring, 11sold, sweepsjuly, 345, 35...   

                                    Representative_Docs  
0     [stock value isnt useful whats interesting fla...  
1     [solar panel roof, car battery solar roof, sol...  
2                                       [nan, nan, nan]  
3     [gspc nflx share rise earnings beat, gspc nflx...  
4                                    [elon, elon, elon]  
...                                                 ...  
2122  [weekly signal issued monday amfor solid move ...  
2123      [go month bynd burger, coming burger, burger]  
2124  [anyone show sale increased localnational gove...  
2125  [go time high believe achieve 4200 share year ...  
2126  [call sweepsjuly 19th expiry, 11sold call area...  

[2127 rows x 5 columns]

Ya que se obtienen un gran número de documentos pertenecientes al tópico -1 (outlier), se reducen los mismos y se actualizan las representaciones de los tópicos obtenidos

In [ ]:
new_topics = topic_model.reduce_outliers(data, topics)
topic_model.update_topics(data, topics=new_topics)

Se obtiene la información general sin tantos outliers

In [ ]:
freq = topic_model.get_topic_info()
freq

Topic  Count                                          Name  \
0        -1    192    -1_oniontsla_tslaentry_illegaltsla_caching   
1         0   2180                   0_solar_roof_solarcity_tile   
2         1    992                         1_nan_nano_ching_yeah   
3         2   1003                     2_gspc_surge_fall_outlook   
4         3    905                  3_elons_elon_stuttering_tapp   
...     ...    ...                                           ...   
2122   2121     18  2121_issued_signal_yesterdaystock_alreadyspy   
2123   2122     31          2122_burger_sandwich_ribt_vegitarian   
2124   2123     54   2123_sticker_restriction_extended_household   
2125   2124     18                2124_4200_dealt_obsolete_yuasa   
2126   2125     54                  2125_expiry_magnet_240c_2425   

                                         Representation  \
0     [oniontsla, tslaentry, illegaltsla, caching, t...   
1     [solar, roof, solarcity, tile, city, panel, ba...   
2                  [nan, nano, ching, yeah, , , , , , ]   
3     [gspc, surge, fall, outlook, nflx, soar, rise,...   
4     [elons, elon, stuttering, tapp, lie, wright, g...   
...                                                 ...   
2122  [issued, signal, yesterdaystock, alreadyspy, t...   
2123  [burger, sandwich, ribt, vegitarian, mash, che...   
2124  [sticker, restriction, extended, household, on...   
2125  [4200, dealt, obsolete, yuasa, mitsubishigs, w...   
2126  [expiry, magnet, 240c, 2425, stepping, 19th, 3...   

                                    Representative_Docs  
0     [stock value isnt useful whats interesting fla...  
1     [solar panel roof, car battery solar roof, sol...  
2                                       [nan, nan, nan]  
3     [gspc nflx share rise earnings beat, gspc nflx...  
4                                    [elon, elon, elon]  
...                                                 ...  
2122  [weekly signal issued monday amfor solid move ...  
2123      [go month bynd burger, coming burger, burger]  
2124  [anyone show sale increased localnational gove...  
2125  [go time high believe achieve 4200 share year ...  
2126  [call sweepsjuly 19th expiry, 11sold call area...  

[2127 rows x 5 columns]

Se extraen los n términos top del tópico más relevante/repetido (0) y la puntuación c-TF-IDF de cada uno

In [ ]:
topic_model.get_topic(0)

[('solar', 0.01760027467952953),
 ('roof', 0.017043836447936053),
 ('solarcity', 0.01638360065103524),
 ('tile', 0.010647335229350924),
 ('city', 0.008134965581446704),
 ('panel', 0.007029349518304793),
 ('bailout', 0.005115056075276034),
 ('installation', 0.004787304330391724),
 ('shingle', 0.004786569358671893),
 ('acquisition', 0.003870052534784978)]

Debido al elevado número de tópicos encontrado (2127) y la limitación computacional, no es posible representar la distancia intertópica inicial

In [ ]:
#DISTANCIA INTERTOPICA INICIAL
topic_model.visualize_topics()

Se guarda el modelo para no perder la clasificación inicial cuando expire la sesión

In [ ]:
topic_model.save('/content/drive/MyDrive/TFG ADE/teslatopicinicial')

Se agrupan los tópicos para obtener las relaciones existentes entre ellos y se visualiza la jerarquía de los 25 más relevantes, con el objetivo de reducir el número de tópicos

In [ ]:
#JERARQUIA INICIAL QUE AYUDA A REDUCIR EL NUMERO DE TOPICOS
topic_model.visualize_hierarchy(top_n_topics=25)

Se vuelve a cargar el modelo

In [ ]:
modelo = BERTopic.load('/content/drive/MyDrive/TFG ADE/teslatopicinicial')

Se van reduciendo el número de tópicos y visualizando la distancia intertópica hasta obtener el número óptimo (20) en el que la distancia es lo suficientemente grande como para que estos sean diferenciables

Se van reduciendo el número de tópicos y visualizando la distancia intertópica hasta obtener el número óptimo (22) en el que la distancia es lo suficientemente grande como para que estos sean diferenciables

In [ ]:
#PRUEBAS TOPIC REDUCTION Y DISTANCIA INTERTOPICA RESULTANTE
modelo.reduce_topics(data, nr_topics=20)
modelo.visualize_topics()

2023-11-14 18:09:16,245 - BERTopic - Reduced number of topics from 22 to 20


De nuevo se obtiene la información general sobre los 20 tópicos resultantes. Se observa que el 1, 17 y 18 corresponden a acumulación de tweets "basura", por lo que no se tendrán en cuenta para el posterior análisis

In [ ]:
modelo.get_topic_info()

Topic   Count                                             Name  \
0      -1     192       -1_oniontsla_tslaentry_illegaltsla_caching   
1       0  187446                         0_trade_stock_nflx_model   
2       1     992                            1_nan_nano_ching_yeah   
3       2     772                2_cnbc_cramer_cnbcfastmoney_stock   
4       3     627            3_bagholderquotes_tent_baggie_gifford   
5       4     281               4_lidar_vaporware_rendering_render   
6       5     200                          5_lotto_call_put_friday   
7       6     133            6_q32019_q3earnings_qtrresearch_model   
8       7     110                          7_vwap_bwaa_bwaaa_break   
9       8     110               8_advice_investment_long_providing   
10      9     107       9_whistleblower_employee_retaliation_fraud   
11     10     106              10_anxiety_sociopath_range_disorder   
12     11      92           11_consolidation_resistance_break_move   
13     12      92              12_allelectric_porsche_model_taycan   
14     13      79           13_timseymour_short_cnbc_cnbcfastmoney   
15     14      68        14_cheer_cheerleader_squeezy_cheerleading   
16     15      67                       15_wash_rinse_washing_leak   
17     16      47  16_virtue_signaling_signalling_virtuesignalling   
18     17      45    17_9999999999_disclaimerpremarket_1000_100000   
19     18      32                      18_let_goooo_goooooo_gooooo   

                                       Representation  \
0   [oniontsla, tslaentry, illegaltsla, caching, t...   
1   [trade, stock, nflx, model, nvda, short, elon,...   
2                [nan, nano, ching, yeah, , , , , , ]   
3   [cnbc, cramer, cnbcfastmoney, stock, believer,...   
4   [bagholderquotes, tent, baggie, gifford, bagho...   
5   [lidar, vaporware, rendering, render, product,...   
6   [lotto, call, put, friday, play, ticket, bough...   
7   [q32019, q3earnings, qtrresearch, model, produ...   
8   [vwap, bwaa, bwaaa, break, entry, short, long,...   
9   [advice, investment, long, providing, give, sh...   
10  [whistleblower, employee, retaliation, fraud, ...   
11  [anxiety, sociopath, range, disorder, psychopa...   
12  [consolidation, resistance, break, move, trend...   
13  [allelectric, porsche, model, taycan, truck, p...   
14  [timseymour, short, cnbc, cnbcfastmoney, posit...   
15  [cheer, cheerleader, squeezy, cheerleading, oc...   
16  [wash, rinse, washing, leak, water, repeat, ca...   
17  [virtue, signaling, signalling, virtuesignalli...   
18  [9999999999, disclaimerpremarket, 1000, 100000...   
19  [let, goooo, goooooo, gooooo, gooo, gooooooooo...   

                                  Representative_Docs  
0                   [oniontsla, oniontsla, oniontsla]  
1   [dont miss next free option trade sign daily f...  
2                                     [nan, nan, nan]  
3                     [problem cnbc, cnbc time, cnbc]  
4   [built tent, bagholderquotes, good bagholderqu...  
5   [committed lidar useless dumping lidar cheap s...  
6                         [lotto lotto, lotto, lotto]  
7   [opportunity increase demand q3earnings q32019...  
8   [covered vwap held vwap bounce, vwap call, vwa...  
9      [free advice worth, give advice, great advice]  
10  [year later whistleblower ignored, national wh...  
11  [range anxiety affect driver driver, soon reme...  
12  [consolidation higher, break consolidation, mo...  
13  [ford show allelectric f150 prototype, capture...  
14  [timseymour short cover, good timseymour, tims...  
15  [input cheer, higher first cheer, picked share...  
16  [rinse wash repeat, someone forgot tell tslas ...  
17  [virtue signaling sucka, virtue signalling cro...  
18  [1000 tell 1000, 1000, quality 19999999999 999...  
19                  [let goooo, let goooo, let goooo]

Se guarda el modelo con los 20 tópicos resultantes para no perderlo cuando expire la sesión

In [ ]:
modelo.save('/content/drive/MyDrive/TFG ADE/teslatopic20')

Se obtene la información sobre los documentos que sirvieron como base para la categorización de gtópicos, incluyendo los documentos en sí, sus respectivos temas, el nombre de cada tema, las n palabras principales de cada tema, si es un documento representativo y la probabilidad de agrupación si se trata del modelo de agrupación

In [ ]:
modelo.get_document_info(data)

Document  Topic  \
0       patent portfolio company combined self driving...      0   
1       joint development driving system high automati...      0   
2       suck never heard dealer able least find equiva...      0   
3       elonmusk autopilot almost another twice servic...      0   
4                                             paid spread      0   
...                                                   ...    ...   
191593  getting wasted thanks nhtsagov approved autopi...      0   
191594                                   interesting dude      0   
191595  2020 brought selfawareness autopilot asshole s...      0   
191596  2020 start tweeting positive news mcep every n...      0   
191597             patiently waiting twitter sitter tweet      0   

                            Name  \
0       0_trade_stock_nflx_model   
1       0_trade_stock_nflx_model   
2       0_trade_stock_nflx_model   
3       0_trade_stock_nflx_model   
4       0_trade_stock_nflx_model   
...                          ...   
191593  0_trade_stock_nflx_model   
191594  0_trade_stock_nflx_model   
191595  0_trade_stock_nflx_model   
191596  0_trade_stock_nflx_model   
191597  0_trade_stock_nflx_model   

                                           Representation  \
0       [trade, stock, nflx, model, nvda, short, elon,...   
1       [trade, stock, nflx, model, nvda, short, elon,...   
2       [trade, stock, nflx, model, nvda, short, elon,...   
3       [trade, stock, nflx, model, nvda, short, elon,...   
4       [trade, stock, nflx, model, nvda, short, elon,...   
...                                                   ...   
191593  [trade, stock, nflx, model, nvda, short, elon,...   
191594  [trade, stock, nflx, model, nvda, short, elon,...   
191595  [trade, stock, nflx, model, nvda, short, elon,...   
191596  [trade, stock, nflx, model, nvda, short, elon,...   
191597  [trade, stock, nflx, model, nvda, short, elon,...   

                                      Representative_Docs  \
0       [dont miss next free option trade sign daily f...   
1       [dont miss next free option trade sign daily f...   
2       [dont miss next free option trade sign daily f...   
3       [dont miss next free option trade sign daily f...   
4       [dont miss next free option trade sign daily f...   
...                                                   ...   
191593  [dont miss next free option trade sign daily f...   
191594  [dont miss next free option trade sign daily f...   
191595  [dont miss next free option trade sign daily f...   
191596  [dont miss next free option trade sign daily f...   
191597  [dont miss next free option trade sign daily f...   

                                              Top_n_words  Probability  \
0       trade - stock - nflx - model - nvda - short - ...     0.000000   
1       trade - stock - nflx - model - nvda - short - ...     1.000000   
2       trade - stock - nflx - model - nvda - short - ...     0.738882   
3       trade - stock - nflx - model - nvda - short - ...     0.000000   
4       trade - stock - nflx - model - nvda - short - ...     0.000000   
...                                                   ...          ...   
191593  trade - stock - nflx - model - nvda - short - ...     0.879340   
191594  trade - stock - nflx - model - nvda - short - ...     1.000000   
191595  trade - stock - nflx - model - nvda - short - ...     0.000000   
191596  trade - stock - nflx - model - nvda - short - ...     0.000000   
191597  trade - stock - nflx - model - nvda - short - ...     0.000000   

        Representative_document  
0                         False  
1                         False  
2                         False  
3                         False  
4                         False  
...                         ...  
191593                    False  
191594                    False  
191595                    False  
191596                    False  
191597                    False  

[191598 rows x 8 columns]

Se actualizan las etiquetas estándar de los tópicos resultantes asignadas por el modelo a unas etiquetas custom que faciliten la interpretabilidad de los mismos

In [ ]:
#custom labels
labels_tesla={0: "Inversiones y comparativa entre empresas",
        2: "Análisis CNBC",
        3: "Baillie Gifford",
        4: "Conducción autónoma",
        5: "Lotería",
        6:"Ganancias Q3 2019",
              7:"Comparativa BorgWarner",
              8:"Consejos de inversión",
              9:"Denuncias y acusaciones",
              10:"Range Anxiety",
              11:"Consolidación de acciones",
              12:"Coches eléctricos",
              13:"Especulación CNBC",
              14:"Animadores",
              15:"Lavado de coches",
              16:"Postureo ético"}
modelo.set_topic_labels(labels_tesla)

Se representan en gráficos de barras los 16 tópicos ordenados por relevancia, en los cuales se obtienen los términos más repetidos por tópico y su puntuación c-TF-IDF

In [ ]:
modelo.visualize_barchart(top_n_topics=15,topics=[0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16], custom_labels=True, width=400)

Se representa la matriz de similitud entre los 10 primeros tópicos más relevantes, para comprobar si son lo suficientemente distintos (valores de similitud menor o igual a 0.5)

In [ ]:
modelo.visualize_heatmap(n_clusters=5, topics=[0,2,3,4,5,6,7,8,9,10],custom_labels=True, width=1200,height=1000)

Se añaden las columnas con la categorización por tópico (número y etiqueta custom) y las palabras clave de cada uno en el dataset

In [ ]:
data_tesla['Tópico']=modelo.get_document_info(data_tesla['body'])['Topic']
data_tesla['Categoría']=modelo.get_document_info(data_tesla['body'])['CustomName']
data_tesla['Palabras clave']=modelo.get_document_info(data_tesla['body'])['Top_n_words']

Se descartan los tweets que tratan sobre el tópico -1 (outlier), y los que acumulan tweets "basura" (1,17,18)

In [ ]:
#se descartan los tweets con el outlier y los tópicos con tweets basura
data_tesla.drop(data_tesla[(data_tesla['Tópico'] ==-1)].index, inplace=True)
data_tesla.drop(data_tesla[(data_tesla['Tópico'] ==1)].index, inplace=True)
data_tesla.drop(data_tesla[(data_tesla['Tópico'] ==17)].index, inplace=True)
data_tesla.drop(data_tesla[(data_tesla['Tópico'] ==18)].index, inplace=True)


Se guarda el dataset con las nuevas columnas

In [ ]:
#se guarda para el analisis de sentimientos
data_tesla.to_csv('/content/drive/MyDrive/TFG ADE/TeslaData20Topics.csv')

In [ ]:
data_tesla.head(50)

tweet_id            post_date  \
0   1134610603715309569  2019-06-01 00:00:22   
1   1134610685084688385  2019-06-01 00:00:42   
2   1134610693133611011  2019-06-01 00:00:43   
3   1134610801648689153  2019-06-01 00:01:09   
4   1134610879994040320  2019-06-01 00:01:28   
5   1134610946545127424  2019-06-01 00:01:44   
6   1134611072411942912  2019-06-01 00:02:14   
7   1134611235972849664  2019-06-01 00:02:53   
8   1134611314452705280  2019-06-01 00:03:12   
9   1134611457981784064  2019-06-01 00:03:46   
10  1134611628975169537  2019-06-01 00:04:27   
11  1134611845669691392  2019-06-01 00:05:18   
12  1134611879123308544  2019-06-01 00:05:26   
13  1134612054021693440  2019-06-01 00:06:08   
14  1134612150486523905  2019-06-01 00:06:31   
15  1134612287011065857  2019-06-01 00:07:03   
16  1134612431068631040  2019-06-01 00:07:38   
17  1134612734056771591  2019-06-01 00:08:50   
18  1134616881610887168  2019-06-01 00:25:19   
19  1134617480305860608  2019-06-01 00:27:42   
20  1134618093903327232  2019-06-01 00:30:08   
21  1134618511240835073  2019-06-01 00:31:47   
22  1134618904003842048  2019-06-01 00:33:21   
23  1134619016885133312  2019-06-01 00:33:48   
24  1134619400428908545  2019-06-01 00:35:19   
25  1134619481811144704  2019-06-01 00:35:39   
26  1134619570508062725  2019-06-01 00:36:00   
27  1134620307313086469  2019-06-01 00:38:56   
28  1134621183188635648  2019-06-01 00:42:24   
29  1134621896094420993  2019-06-01 00:45:14   
30  1134622406025158656  2019-06-01 00:47:16   
31  1134622579732504578  2019-06-01 00:47:57   
32  1134622679632371714  2019-06-01 00:48:21   
33  1134623498083737605  2019-06-01 00:51:36   
34  1134623595253161984  2019-06-01 00:52:00   
35  1134623639989604352  2019-06-01 00:52:10   
36  1134624351070883840  2019-06-01 00:55:00   
37  1134624363729248256  2019-06-01 00:55:03   
38  1134624504754257920  2019-06-01 00:55:36   
39  1134624724653228032  2019-06-01 00:56:29   
40  1134624819742330880  2019-06-01 00:56:51   
41  1134625294248239104  2019-06-01 00:58:45   
42  1134625372178399233  2019-06-01 00:59:03   
43  1134625694930079744  2019-06-01 01:00:20   
44  1134625727377215490  2019-06-01 01:00:28   
45  1134625837213372416  2019-06-01 01:00:54   
46  1134625950212059136  2019-06-01 01:01:21   
47  1134627145446240257  2019-06-01 01:06:06   
48  1134628100300509186  2019-06-01 01:09:54   
49  1134632160781053952  2019-06-01 01:26:02   

                                                 body  Tópico  \
0   patent portfolio company combined self driving...       0   
1   joint development driving system high automati...       0   
2   suck never heard dealer able least find equiva...       0   
3   elonmusk autopilot almost another twice servic...       0   
4                                         paid spread       0   
5   really elon fund recently hopefully spend afte...       0   
6   doesnt sound safe recommendation service cente...       0   
7   instead getting money long though much want ac...       0   
8   great problemunfortunately customer report pro...       0   
9                            statement reluctant data       0   
10  fraudulent paid combined million fine illegal ...       0   
11  unfortunately issueteslaserviceissues whereare...       0   
12      harem delivered summon although longer afford       0   
13  wasnt expected guided changed price configurat...       0   
14                                  long gladly money       0   
15  youre patient individual however solar company...       0   
16  insurance must approached filing waste money c...       0   
17  bummer almost theyve shut solar division least...       0   
18  false projectionsinvestor appearance hypergrow...       0   
19  disingenuous connected vested interest custome...       0   
20            active trade free trade guide sbux nvda       0   
21                              pivotalcapital filing       0   
22                        apparently short conflicted       0   
23  mo

Se obtiene la variación frecuencia por tópico a lo largo de la franja temporal de estudio

In [ ]:
#cambiar a mes
data_tesla['post_date_only']= pd.to_datetime(data_tesla['post_date']).dt.date
modelo = BERTopic.load('/content/drive/MyDrive/TFG ADE/teslatopic20')
topics_over_time = modelo.topics_over_time(data_tesla['body'].astype(str), data_tesla['post_date_only'])

In [ ]:
modelo.visualize_topics_over_time(topics_over_time, topics=[0, 2, 3, 4,5], custom_labels=True, width=1500)

In [ ]:
modelo.visualize_topics_over_time(topics_over_time, topics=[6,7, 8, 9, 10], custom_labels=True, width=1500)

In [ ]:
modelo.visualize_topics_over_time(topics_over_time, topics=[11,12,13, 14, 15, 16], custom_labels=True, width=1500)

Se exporta la evolución temporal por tópico para representarla en otra herramienta

In [ ]:
# evolucion temporal topicos
data_tesla['post_date'] = pd.to_datetime(data_tesla['post_date'])
topics = [0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16]

for topic in topics:
  topic_tweets = data_tesla[data_tesla['Tópico'] == topic]

  # Group the data by day and count the number of tweets per day
  t_per_day = topic_tweets.resample('D', on='post_date').count()
  columns_to_delete = ['body',  'Categoría', 'Palabras clave', 'Unnamed: 0', 'tweet_id']
  t_per_day = t_per_day.drop(columns=columns_to_delete)
  t_per_day = t_per_day.rename(columns={'Tópico': f'Tweets diarios sobre tópico {topic}'})
  t_per_day.to_csv(f'drive/MyDrive/TFG ADE/tesla{topic}.csv')